cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,501 @@
+	if withNumSegments {
+		pointerNumSegments = int64(numSegments)
+	}
+
+	metadata, err := proto.Marshal(&pb.StreamMeta{
+		NumberOfSegments: pointerNumSegments,
+	})
+	require.NoError(t, err)
+
+	raw, err := metainfo.CreatePath(ctx, *projectID, -1, []byte(bucketName), []byte(objectID))
+	require.NoError(t, err)
+
+	references = append(references, segmentRef{
+		path: metainfo.ScopedPath{
+			ProjectID:           *projectID,
+			ProjectIDString:     projectIDString,
+			BucketName:          bucketName,
+			Segment:             "l",
+			EncryptedObjectPath: encryptedPath,
+			Raw:                 raw,
+		},
+		pointer: &pb.Pointer{
+			Type:     pointerType,
+			Metadata: metadata,
+		},
+	})
+
+	return encryptedPath, references
+}
+
+type testdataObjects struct {
+	// expectations
+	expectedNumSegments    int
+	expectedInlineSegments int
+	expectedRemoteSegments int
+	expectedObjects        bucketsObjects
+
+	// data used for calling processSegment
+	objSegments []segmentRef
+	projectID   *uuid.UUID
+}
+
+// generateTestdataObjects generate a testdataObjecst with a random number of
+// segments of a random number of objects and buckets but under the same
+// project.
+//
+// When withoutLastSegment is true, there will be objects without last segment,
+// otherwise all of them will have a last segment.
+//
+// When withMoreThanMaxNumSegments is true, there will be objects with more
+// segments than the maxNumOfSegments, otherwise all of them will have less or
+// equal than it.
+// nolint:golint
+func generateTestdataObjects(
+	t *testing.T, ctx context.Context, withoutLastSegment bool, withMoreThanMaxNumSegments bool,
+) testdataObjects {
+	t.Helper()
+
+	var (
+		testdata = testdataObjects{
+			expectedObjects: make(bucketsObjects),
+		}
+		bucketName                      = "0"
+		numObjs                         = rand.Intn(10) + 2
+		projID                          = testrand.UUID()
+		withoutLastSegmentCount         = 0
+		withMoreThanMaxNumSegmentsCount = 0
+		numMaxGeneratedSegments         = 10
+	)
+
+	if withMoreThanMaxNumSegments {
+		numMaxGeneratedSegments = 100
+	}
+
+	testdata.projectID = &projID
+
+	for i := 0; i < numObjs; i++ {
+		var (
+			inline          = (rand.Int() % 2) == 0
+			withNumSegments = (rand.Int() % 2) == 0
+			numSegments     = rand.Intn(numMaxGeneratedSegments) + 2
+		)
+
+		if numSegments > (int(maxNumOfSegments) + 1) {
+			withMoreThanMaxNumSegmentsCount++
+		}
+
+		// If withMoreThanMaxNumSegments is true and all the objects created in all
+		// the previous iterations have less or equal than maximum number of segments
+		// and this is the last iteration then force that the object crated in this
+		// iteration has more segments than the maximum.
+		if withMoreThanMaxNumSegments &&
+			withMoreThanMaxNumSegmentsCount == 0 &&
+			i == (numObjs-1) {
+			numSegments += int(maxNumOfSegments)
+			withMoreThanMaxNumSegmentsCount++
+		}
+
+		if rand.Int()%2 == 0 {
+			bucketName = fmt.Sprintf("bucket-%d", i)
+		}
```

Comment:
What's the purpose of this?

Reply:
randomly generate the object in the previously used bucket or in a new bucket, for not being stale to always generate objects in single bucket or always generate buckets with only one object.